semcore/input-number/src/InputNumber.jsx

```diff
@@ -106,51 +180,271 @@ class Value extends Component {
-    this.inputRef.current?.addEventListener('wheel', this.handleWheel);
+    this.valueInputRef.current?.addEventListener('wheel', this.handleWheel);
+
+    const { inputRef, value } = this.asProps;
+
+    if (inputRef.current) {
+      inputRef.current.stepUp = this.stepUp;
+      inputRef.current.stepDown = this.stepDown;
+    }
+
+    if (value !== '') {
+      const { displayValue } = this.valueParser(value, '', '');
+      this.handlers.displayValue(displayValue);
+    }
   }
   componentWillUnmount() {
-    this.inputRef.current?.removeEventListener('wheel', this.handleWheel);
+    this.valueInputRef.current?.removeEventListener('wheel', this.handleWheel);
   }
 
   handleWheel = (event) => {
-    if (event.target !== this.inputRef.current) return;
-    if (document.activeElement !== this.inputRef.current) return;
+    if (event.target !== this.valueInputRef.current) return;
+    if (document.activeElement !== this.valueInputRef.current) return;
     event.preventDefault();
     if (event.wheelDelta > 0) {
-      this.asProps.increment(event);
+      this.stepUp(event);
     } else if (event.wheelDelta < 0) {
-      this.asProps.decrement(event);
+      this.stepDown(event);
+    }
+  };
+
+  handleChange = (event) => {
+    const value = event.currentTarget.value;
+    const digits = /[0-9,.-]+/.test(value);
+
+    if (digits || value === '') {
+      this.handlers.value(value, event);
+    }
+
+    return false;
+  };
+
+  handleKeyUp = (event) => {
+    if (event.key === 'Shift') {
+      this.cursorPosition = -1;
+    }
+
+    const element = event.currentTarget;
+
+    element.role = 'input';
+  };
+
+  handleKeyDown = (event) => {
+    const element = event.currentTarget;
+    const value = element.value;
+    const length = value.length;
+
+    // we need this dirty hack for screen readers, because they couldn't read full value in input after adding there ','.
+    // so, we change role to `region` here and back to `input` in handleKeyUp
+    element.role = 'region';
+
+    if (
+      element.selectionStart !== length &&
+      (event.key === 'Backspace' ||
+        event.key === this.separatorDecimal ||
+        ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'].includes(event.key))
+    ) {
+      const afterSelection = value.slice(element.selectionEnd);
+
+      requestAnimationFrame(() => {
+        const newValue = this.state.displayValue;
+        const index = newValue.lastIndexOf(afterSelection);
+        const selectionStart = index;
+        const selectionEnd = index;
+
+        element.setSelectionRange(selectionStart, selectionEnd);
+      });
+    }
+
+    // For correct moving cursor with skip separatorThousands.
+    // Examples:
+    // - Press ArrowLeft: `12,3|4 -> 12|,34`
+    // - Press ArrowRight: `1|,55 -> 1,5|5`
+    const cursorIndex = 2;
+
+    if (event.shiftKey && this.cursorPosition === -1) {
+      this.cursorPosition = element.selectionStart;
+    }
+
+    if (this.cursorPosition !== -1) {
```

Comment:
This tree of ifs looks is not easy to read. Could you somehow refactor it into more compact way?

Reply:
refactored a little bit